controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -1252,3 +1252,115 @@ def test_identity_probe(self):
+
+        self.context = utils.FakeContext()
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_success(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+
+        self.servicer.ValidateVolumeCapabilities(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.OK)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_empty_id(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_id = ""
+
+        res = self.servicer.ValidateVolumeCapabilities(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("volume id" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_wrong_secrets(self, storage_agent):
+        self._test_create_object_with_wrong_secrets(storage_agent)
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_unsupported_capabilities(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_capabilities[0].access_mode.mode = 999
+
+        res = self.servicer.ValidateVolumeCapabilities(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("unsupported access mode" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_no_capabilities(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_capabilities = {}
+
+        res = self.servicer.ValidateVolumeCapabilities(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.INVALID_ARGUMENT)
+        self.assertTrue("not set" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))
+
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_validate_volume_cap_with_bad_id(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.request.volume_id = "wwn1"
+
+        res = self.servicer.ValidateVolumeCapabilities(self.request, self.context)
+
+        self.assertEqual(self.context.code, grpc.StatusCode.NOT_FOUND)
+        self.assertTrue("id format" in self.context.details)
+        self.assertEqual(res, self.get_create_object_response_method()(message=self.context.details))

```

Comment:
would it be possible to wrap these 3 lines? e.g.
```
def assertResponse(self, response, expected_status_code, expected_details_substring):
```

Reply:
done